Mathlib/Analysis/CStarAlgebra/CStarVec.lean

```diff
@@ -0,0 +1,329 @@
+    · rintro rfl; simp only [inner, CStarVec.zero_apply, star_zero, mul_zero, Finset.sum_const_zero]
+  inner_op_smul_right {a} {x} {y} := by
+    simp only [inner, smul_apply, smul_eq_mul, mul_assoc]
+    apply Eq.symm
+    simp [Finset.sum_mul, mul_assoc]
+  inner_smul_right_complex {z} {x} {y} := by
+    simp only [inner_eq_sum, smul_apply, smul_mul_assoc]
+    apply Eq.symm
+    simp only [Finset.smul_sum, mul_smul_comm]
+  star_inner x y := by simp only [inner, star_sum, star_mul, star_star]
+  norm_eq_sqrt_norm_inner_self v := rfl
+
+lemma norm_entry_le_norm [DecidableEq n] [CompleteSpace A] {v : CStarVec n A} {i : n} :
+    ‖v i‖ ≤ ‖v‖ := by
+  by_cases htriv : 0 < ‖v i‖
+  · have hmain := calc ‖v i‖ * ‖v i‖ = ‖star (v i) * v i‖  := CStarRing.norm_star_mul_self.symm
+      _ = ‖⟪ofFun (Pi.single i (v i)), v⟫_A‖ := by rw [inner_single_eq_entry]
+      _ ≤ ‖ofFun (Pi.single i (v i))‖ * ‖v‖ :=
+            CStarModule.norm_inner_le (E := CStarVec n A)
+      _ = ‖v i‖ * ‖v‖ := by rw [norm_single]
+    rwa [mul_le_mul_left htriv] at hmain
+  · push_neg at htriv
+    simp only [norm_le_zero_iff] at htriv
+    simp only [htriv, _root_.norm_zero]
+    exact CStarModule.norm_nonneg
+
+end inner
+
+section TopologyAux
+/-
+## Replacing the uniformity and bornology
+
+Note that while the norm that we have defined on `CStarVec n A` induces the product uniformity,
+it is not defeq to `Pi.uniformSpace`. In this section, we show that the norm indeed does induce
+the product topology and use this fact to properly set up the `NormedAddCommGroup (CStarVec n A)`
+instance such that the uniformity is still `Pi.uniformSpace` and the bornology is
+`Pi.instBornology`.
+
+To do this, we first temporarily register the "bad" `NormedAddCommGroup (CStarVec n A)` instance
+and show that the map from `ofFun.symm : CStarVec n A → (n → A)` is both Lipschitz and
+Antilipschitz.  We then finally register the `NormedAddCommGroup (CStarVec n A)` instance via
+`NormedAddCommGroup.ofCoreReplaceAll`.
+-/
+
+variable {A : Type*} [NonUnitalNormedRing A] [StarRing A] [CStarRing A] [PartialOrder A]
+  [CompleteSpace A] [StarOrderedRing A] [NormedSpace ℂ A] [StarModule ℂ A] [IsScalarTower ℂ A A]
+  [SMulCommClass ℂ A A]
+
+variable {n : Type*} [Fintype n]
+
+/-- A temporary "bad" instance of `NormedAddCommGroup` that induces a uniformity that is not
+defeq with the product uniformity. -/
+private noncomputable abbrev normedAddCommGroupAux : NormedAddCommGroup (CStarVec n A) :=
+  NormedAddCommGroup.ofCore CStarModule.normedSpaceCore
+
+attribute [local instance] normedAddCommGroupAux
+
+/-- A temporary "bad" instance of `NormedSpace`. -/
+private noncomputable abbrev normedSpaceAux : NormedSpace ℂ (CStarVec n A) :=
+  .ofCore CStarModule.normedSpaceCore
+
+attribute [local instance] normedSpaceAux
+
+open Finset Classical in
+private lemma lipschitzWith_ofFun_symm_aux :
+    LipschitzWith 1 (ofFun.symm : CStarVec n A → n → A) := by
```

Comment:
Are these proofs essentially just copied from the analogous ones for `EuclideanSpace`?

Reply:
No, I mostly looked at `PiLp` for inspiration for this file actually. I followed the same general strategy, but I reproved the Lipschitz/Antilipschitz from scratch.